core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,164 @@
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            checkBytes()
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError) return
+        status = NSStreamStatusClosed
+        source.close()
+        runLoop = null
+        runLoopModes = listOf()
+    }
+
+    override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        try {
+            if (isClosed()) throw IOException("Underlying source is closed.")
+            if (status != NSStreamStatusOpen && status != NSStreamStatusAtEnd) {
+                return -1
+            }
+            if (source.exhausted()) {
+                status = NSStreamStatusAtEnd
+                return 0
+            }
+            if (buffer == null) return -1
+
+            status = NSStreamStatusReading
+            val toRead = minOf(maxLength.toInt(), source.buffer.size).toInt()
+            val read = source.buffer.readAtMostTo(buffer, toRead).convert<NSInteger>()
+            status = NSStreamStatusOpen
+            checkBytes()
+            return read
+        } catch (e: Exception) {
+            error = e.toNSError()
+            return -1
+        }
+    }
+
+    override fun getBuffer(buffer: CPointer<CPointerVar<uint8_tVar>>?, length: CPointer<NSUIntegerVar>?) = false
+
+    override fun hasBytesAvailable() = !source.exhausted()
+
+    override fun propertyForKey(key: NSStreamPropertyKey): Any? = null
+
+    override fun setProperty(property: Any?, forKey: NSStreamPropertyKey) = false
+
+    private var _delegate = WeakReference<NSStreamDelegateProtocol>(this)
+    private var runLoop: NSRunLoop? = null
+    private var runLoopModes = listOf<NSRunLoopMode>()
+
+    private fun postEvent(event: NSStreamEvent) {
+        val runLoop = runLoop ?: return
+        runLoop.performInModes(runLoopModes) {
+            if (runLoop == this.runLoop) {
+                delegate?.stream(this, event)
+            }
+        }
+    }
+
+    private fun checkBytes(sendEndEvent: Boolean = true) {
+        val runLoop = runLoop ?: return
+        runLoop.performInModes(runLoopModes) {
+            if (runLoop != this.runLoop) return@performInModes
+            try {
+                if (source.exhausted()) {
+                    if (sendEndEvent) delegate?.stream(this, NSStreamEventEndEncountered)
+                    val timer = NSTimer.timerWithTimeInterval(0.1, false) {
+                        checkBytes(sendEndEvent = false)
+                    }
+                    runLoopModes.forEach { mode ->
+                        runLoop.addTimer(timer, mode)
+                    }
+                } else {
+                    status = NSStreamStatusOpen
+                    delegate?.stream(this, NSStreamEventHasBytesAvailable)
+                }
+            } catch (e: IllegalStateException) {
+                // ignore closed
```

Comment:
My intention was to avoid crashing with a possible race condition if the stream was closed on another thread immediately after the run loop thread checked the closed state. I would think typical usage when using the run loop would be to close the stream on the run loop thread after receiving the `NSStreamEventEndEncountered` event, but there might be a reason a consumer would prematurely close the stream and wouldn't want to crash here if that were the case.

Reply:
Fixed: https://github.com/Kotlin/kotlinx-io/pull/174/commits/ed5902ec5edf09000c06f7c0609b1ecca519295b